### DISCUSSION TITLE
Хранить файлы пользователя на внешнем ресурсе

### DISCUSSION THREAD

**Discussion by vldekov** [totumonline/totum-ru-issues-and-discussions#340] [Как это сделать?] [answered]
Есть возможность настроить приложение для хранения файлов пользователя на внешнем ресурсе ( яндекс облако, свой сервер, др. )?

**Comment by ttm-alexeypolunin** [comment]
Если на уровне сервера — примонтируйте папку любым доступным вам способом.

Если резервные копии отправлять в S3 или другие подобные хранилища — ставите их тул на сервер и используете функцию `execSsh`.

Если на уровне поля Файл из интерфейса – функционал в списке, сроки реализации неизвестны. Если очень необходимо ускорить — 1 000 € с документами.

**Reply by vldekov** [reply]
Есть инструкция как примонтировать папку яндекс диска для хранения на уровне поля Файл?

**Reply by ttm-alexeypolunin** [reply]
Функционал в списке, сроки реализации неизвестны — но на самом деле врядли мы будем делать Яндекс.

Если на уровне linux — 

Из google 1: https://www.kost.su/%D0%BC%D0%BE%D0%BD%D1%82%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5-yandexdisk-%D0%BD%D0%B0-linux/

Из google 2: https://blog.xenot.ru/connect-yadisk-ubuntu-webdav-davfs2.fuck 

👆 я не проверял (нашел в интернете)

**Инструкция в закрепе**

**Важно:** надо понимать, что срок ответа вашего сервера удлиняется на срок ответа сервера яндекса (я тот не торопится). Особенно это будет доставлять на загрузке превью картинок в строчной части таблицы.


**Официал от яндекса:** https://yandex.ru/support/disk-desktop-linux/ — их утилита хранит файлы в том числе на локальной машине и синхронизирует их. Те нет задержек по чтению, но есть расход HDD на сервере.

**Reply by vldekov** [reply]
Большое спасибо, будем пробовать!

**Comment by ttm-alexeypolunin** [comment, marked as answer]
### UPD: Вообще Яндекс против использования webDav. Вопрос закрыт — точно не работает нормально. [Тред на stackoverflow](https://ru.stackoverflow.com/questions/1043523/%D0%9F%D0%BE%D1%81%D0%BB%D0%B5-%D0%B7%D0%B0%D0%B3%D1%80%D1%83%D0%B7%D0%BA%D0%B8-%D1%84%D0%B0%D0%B9%D0%BB%D0%B0-%D0%BD%D0%B0-%D0%AF%D0%BD%D0%B4%D0%B5%D0%BA%D1%81-%D0%94%D0%B8%D1%81%D0%BA-%D0%BF%D0%BE-webdav-%D0%BF%D1%80%D0%BE%D0%B8%D1%81%D1%85%D0%BE%D0%B4%D1%8F%D1%82-%D0%BF%D0%B0%D1%83%D0%B7%D1%8B-%D0%BF%D0%BE-60-%D1%81%D0%B5%D0%BA%D1%83%D0%BD%D0%B4-%D0%BD%D0%B0-%D0%BA)

WebDAV не поддерживает символические ссылки, поэтому что бы подключить Яндекс Диск вместо стандартной папки с файлами в Totum надо:

 **Важное замечание:** `Защищенный доступ к файлам` работает по **другому пути!** Он находится вне пределов папки `fls` и его надо монтировать отдельно. Я не проверял.
 
 **Второе важное замечание:  скорость чтения/записи реально очень плохая!**
 
  **Третье важное замечание:  даже с параметром `_netdev` в `fstab` сервак как-то тяжело поднимается**

Как раширять место на серваке по нормальному — в [первом комментарии.](https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/340#discussioncomment-5455234)

Особенно с превью картинок:

<img width="271" alt="Снимок экрана 2023-03-28 в 17 37 17" src="https://user-images.githubusercontent.com/55755565/228274289-3e0efe03-2443-47e8-b0bf-7d75b6ee5266.png">

_Они загрузились, но через минуту._


#### Для сохранения на Яндекс-диск:

1. Откройте [Яндекс ID](https://passport.yandex.ru/profile).
2. В разделе **Безопасность** — **Пароли и авторизация** нажмите **Включить пароли приложений.** Затем введите пароль, чтобы подтвердить действие. Если у вас уже включены пароли приложений, пропустите этот шаг.
3. Нажмите Создать пароль приложения в типе приложения Файлы WebDAV.
4. Введите название пароля (например, «Pass for WebDAV»).
5. Нажмите Создать.
6. Скопируйте показанный пароль.


От `root`:

Устанавливаем `davfs2`:

```
apt install davfs2
```

На вопрос отвечаем `YES`.

Переименуйте папку вашего домена в `/home/totum/totum-mit/https/fls/YOUR_DOMAIN_old`

Создаем точку монтирования:

```
mkdir /home/totum/totum-mit/https/fls/YOUR_DOMAIN
```

Монтируем:

```
mount -t davfs https://webdav.yandex.ru /home/totum/totum-mit/https/fls/YOUR_DOMAIN
```

Вводим логин `user@yandex.ru` `[Enter]`.

Пароль полученный на предыдущем шаге `pass` `[Enter]`.

Проверяем монтирование:

```
ls -lh /home/totum/totum-mit/https/fls/YOUR_DOMAIN
```

**Не копируйте пока сюда старые файлы тк на текущий момент монтирование проведено от `root` и `totum` не будет иметь доступа с перемещенный файлам!**

Создаем секрет:

```
nano /etc/davfs2/secrets 
```

В конец файла вставляем (заменить `YOUR_YSERNAME`, `YOUR_PASS`):

```
https://webdav.yandex.ru YOUR_YSERNAME YOUR_PASS
```

Открываем `fstab`:

```
nano /etc/fstab
```

Вставляем в конец:

*Что бы монтировать от пользователя `totum`:*

```
https://webdav.yandex.ru   /home/totum/totum-mit/https/fls/YOUR_DOMAIN   davfs   rw,user,_netdev,uid=totum,gid=totum    0   0
```

Перезагружаем:

```
reboot
```

Проверяем:

```
ls -lh /home/totum/totum-mit/https/fls/YOUR_DOMAIN
```

*Первое подключение (да и не первое тоже) долго может проходить!*

Теперь можно скопировать файлы из старой папки.

**Reply by ttm-alexeypolunin** [reply]
Те логика какая — хотите сэкономить на диске, все будет печально работать.

Что бы летало, надо `nvme` диски и отсутсвие коннекта через `https` куда-то. Поэтому если нужно допместо, подключайте диски в ЦОДе.

> Чего нам здесь нужно будет сделать — это дробление для поля Файл внутри `fls` (точнее внутри защищенной папки, не будем для открытых файлов это делать) на папки по полю. И внутри папки поля еще на интервалы `id`. Запишем в задачи. Если нужно сейчас: ~1К евро это где-то по времени.

`webDAV` годится для бекапирования в фоне например. Монтируете и по расписанию через `rsync/rclone` отправляете туда файлы. Для оперативного чтения не подходит. По `rsync/rclone` инструкций от нас не будет.

Для синхронизации перспективнее наверное (но это не точно) синхронизировать папку через родную утилиту от Яндекса: https://yandex.ru/support/disk-desktop-linux/

**Reply by ttm-alexeypolunin** [reply]
UPD: Вообще Яндекс против использования webDav. Вопрос закрыт — точно не работает нормально. [Тред на stackoverflow](https://ru.stackoverflow.com/questions/1043523/%D0%9F%D0%BE%D1%81%D0%BB%D0%B5-%D0%B7%D0%B0%D0%B3%D1%80%D1%83%D0%B7%D0%BA%D0%B8-%D1%84%D0%B0%D0%B9%D0%BB%D0%B0-%D0%BD%D0%B0-%D0%AF%D0%BD%D0%B4%D0%B5%D0%BA%D1%81-%D0%94%D0%B8%D1%81%D0%BA-%D0%BF%D0%BE-webdav-%D0%BF%D1%80%D0%BE%D0%B8%D1%81%D1%85%D0%BE%D0%B4%D1%8F%D1%82-%D0%BF%D0%B0%D1%83%D0%B7%D1%8B-%D0%BF%D0%BE-60-%D1%81%D0%B5%D0%BA%D1%83%D0%BD%D0%B4-%D0%BD%D0%B0-%D0%BA)
